universum/modules/output/base_output.py

```diff
@@ -3,46 +3,130 @@
         raise NotImplementedError
 
     def log_error(self, description: str) -> None:
+        """
+        Print an error message to the output. Error may use different colors or other formatting.
+
+        :param description: the detailed error message
+        """
         raise NotImplementedError
 
     def log_external_command(self, command: str) -> None:
+        """
+        Print the start of the external command with parameters.
+
+        :param command: external command with parameters
+        """
         raise NotImplementedError
 
     def log_stdout(self, line: str) -> None:
+        """
+        Print one line of the standard output of the previously launched external command.
+
+        :param line: the line to print
+        """
         raise NotImplementedError
 
     def log_stderr(self, line: str) -> None:
+        """
+        Print one line of the standard error output of the previously launched external command.
+
+        :param line: the line to print
+        """
         raise NotImplementedError
 
     def open_block(self, num_str: str, name: str) -> None:
+        """
+        Start the block of the output. Block is a group of related output lines.
+
+        :param num_str: the string that represents the block number within the hierarchy.
+            Inner blocks are numbered using dot notation.
+            The number string is passed as a separate parameter to allow implementing different formatting for number and
+            name.
+        :param name: the title of the block.
+        """
         raise NotImplementedError
 
     def close_block(self, num_str: str, name: str, status: str) -> None:
+        """
+        End the block of the output.
+
+        :param num_str: the string that represents the block number
+        :param name: the title of the block
+        :param status: the completion status of the block: "Passed" or "Failed"
+        """
         raise NotImplementedError
 
     def log_skipped(self, message: str) -> None:
+        """
+        Print a message related to skipping steps. It could be a warning that further steps are skipped, or a message
+        that the step is skipped.
+
+        :param message: the message to print
+        """
         raise NotImplementedError
 
     def log_summary_step(self, step_title: str, has_children: bool, status: str) -> None:
+        """
+        Print a summary of the step execution at the end of the run.
+
+        :param step_title: the title of the step
+        :param has_children: whether the step has children or not. Typically, if step has children, the status is not
+            included in the summary.
+        :param status: the status of the step: "Passed", "Failed" or "Skipped"
+        """
         raise NotImplementedError
 
     def report_build_problem(self, description: str) -> None:
+        """
+        TeamCity only. Report run problem to the TeamCity via the output. Also fails the build regardless of the process
```

Comment:
Hmm, maybe rename this method to `report_teamcity_build_problem`? The information that this and the next methods are TeamCity-only are quite hidden.

Reply:
I thought about it, but decided not to do it. Reasons are the following:
1. At some point we may support it for more Jenkins and other servers
2. If it is only for TeamCity it should not be part of the common interface at all. But at the moment it doesn't make sense to separate just two methods.